lib/ecto/query.ex

```diff
@@ -9,6 +9,12 @@ defmodule Ecto.SubQuery do
   @type t :: %__MODULE__{}
 end
 
+
+defmodule Ecto.ValuesList do
+  @moduledoc false
+  defstruct [:values, :schema]
+end
```

Comment:
Do we really need a schema? If the maps all have the same keys, can't we just infer the type? This would make the usage significantly easier.

Reply:
how would you handle more complex types like UUID? I don't think that can be inferred since they're just strings

Reply:
Maybe a schema can be given optionally. I also thought about allowing schemaless key to type maps, like in `Ecto.Changeset`.

Reply:
I'd imagine something with `type/2` + type maps (i.e.: `type(^my_list, {:array, %{a: :string}})`, but it probably would be a bigger change. Guess taking schemas in is okay :]

Reply:
I pushed some WIP changes where the schema is actually being used (most of it in ecto_sql, actually)

Reply:
That's pushing concerns to the adapter and we try to avoid that. The best is to normalize everything in the planner.

In any case, the need for a schema was not something I realized we needed in my first review and I am a bit more worried about the feature now but I can't quite put my finger on it yet.

Reply:
I agree it is a bit clunky to pass a schema. I could also explore the suggestion with the optional type.

Reply:
I had to upgrade `def values` in `Ecto.Query` to a macro, but I'm now able to write something like this:

```elixir
list_a = [%{id: 1, name: "Foo"}, %{id: 2, name: "Peter"}]
list_b = [%{id: 1, name: "Bar"}, %{id: 2, name: "Parker"}]
schema = %{id: :integer, name: :string}
from v1 in values(type(list_a, schema)), join: v2 in values(type(list_b, schema)), on: v1.id == v2.id
```

I can also omit the `type/2` call which will generate a map like `%{id: :any, name: :any}` automatically:

```elixir
list_a = [%{id: 1, name: "Foo"}, %{id: 2, name: "Peter"}]
list_b = [%{id: 1, name: "Bar"}, %{id: 2, name: "Parker"}]

from v1 in values(list_a), join: v2 in values(list_b), on: v1.id == v2.id
```

Now the only problem I still need to solve is the offset for the params (`$1`, etc).

Reply:
I'm also open to enforce wrapping the schema map with `{:array, %{...}}` if it helps with consistency.

Reply:
Can we derive types if a list of structs is formed by embedded schema?

Reply:
@fuelen I could see it being done, you'd just need to assume that they all use the same schema (maybe raise otherwise). But I'm not sure how this would be regarding separation of concerns on ecto/adapters.

Reply:
@fuelen I implemented that, also added the function `Ecto.Type.infer_type_from_value/1` which guesses the ecto type from a given value, which works with most primitive types.